File: frontend/4_js/tasks/normal/19.js
/**
 * Проверка сбалансированности скобок
 * Ввод: строка из скобок: `()[]{}`
 * Вывод: `True`, если скобки правильно сбалансированы
 */
function isBalanced(str) {
  const stack = [];
  const pairs = {
    ")": "(",
    "]": "[",
    "}": "{",
  };
  for (let char of str) {
    if (char === "(" || char === "[" || char === "{") {
      stack.push(char);
    } else if (char === ")" || char === "]" || char === "}") {
      if (stack.pop() !== pairs[char]) {
        return false;
      }
    }
  }
  return stack.length === 0;
}
console.log(isBalanced("([{}])")); // true
console.log(isBalanced("([)]))")); // false

File: backend/0_python/tasks/normal/19.py
'''
Проверка сбалансированности скобок
Ввод: строка из скобок: `()[]{}`
Вывод: `True`, если скобки правильно сбалансированы
'''
def is_balanced(s):
  stack = []
  pairs = {')': '(', ']': '[', '}': '{'}
  for char in s:
    if char in '([{':
      stack.append(char)
    elif char in ')]}':
      if not stack or stack[-1] != pairs[char]:
        return False
      stack.pop()
  return len(stack) == 0
s = input('Вве строку со скобками: ')
print(is_balanced(s))
